### PR TITLE
7.0 account invoice unique customer ref

### PR DESCRIPTION
**Pull request by mcassuto**
This PR add 2 modules: 
1/ unique_supplier_invoice_number which set a constraint on (partner_id, supplier_invoice_number) to be unique
2/ account_invoice_unique_customer_ref which set a constraint on (partner_id, name) to be unique, only for out_invoice or out_refund type account_invoice.

### CODE REVIEW

File: account_invoice_unique_customer_ref/account_invoice.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            raise orm.except_orm(_('Error'), 'Cannot check unique name without id.')
+
+        invoice = self.browse(cr, uid, i_id, context=context)
+        invoice_type = invoice.type
+        invoice_partner = invoice.partner_id
+
+        if invoice_type not in ['out_invoice', 'out_refund']:
+            return True
+
+        sr_ids = self.search(cr, uid, [("type", "=", invoice_type), ("partner_id", "=", invoice_partner.id)],
```
Comment (rvalyi) [inline account_invoice_unique_customer_ref/account_invoice.py]: line too long; max 80

File: account_invoice_unique_customer_ref/account_invoice.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
+# -*- encoding: utf-8 -*-
+# #############################################################################
+#
+# OpenERP, Open Source Management Solution
+# This module copyright (C) 2010 - 2014 Savoir-faire Linux
+# (<http://www.savoirfairelinux.com>).
```
Comment (rvalyi) [inline account_invoice_unique_customer_ref/account_invoice.py]: missing 4 spaces at the beginning of these 3 lines

File: account_invoice_unique_customer_ref/account_invoice.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    def copy(self, cr, uid, id, default=None, context=None):
+        default = default or {}
+        default.update({
+            'name': '',
+        })
+        return super(account_invoice, self).copy(cr, uid, id, default, context)
+
+    def _check_unique_name_insensitive(self, cr, uid, ids, context=None):
+        # this function only works with one ids
```
Comment (rvalyi) [inline account_invoice_unique_customer_ref/account_invoice.py]: "one ids" ? -> "one id"  or "an ids array of a single id"

File: unique_supplier_invoice_number/account_invoice.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    def copy(self, cr, uid, id, default=None, context=None):
+        default = default or {}
+        default.update({
+            'supplier_invoice_number': '',
+        })
+        return super(account_invoice, self).copy(cr, uid, id, default, context)
+
+    def _check_unique_supplier_invoice_number_insensitive(self, cr, uid, ids,
+                                                          context=None):
+        # this function only works with one ids
```
Comment (rvalyi) [inline unique_supplier_invoice_number/account_invoice.py]: "one ids" -> same remark

**Comment by rvalyi** [comment]
Hello I made some non exhaustive comments.

Also, are we accepting a company logo in OCA modules? I'm all in favor that modules authors/companies get the visibility that fuel the work, but I think till now we were not accepting company logos. If we do, we should see how we do it (I'm not against but only if it is highly moderated).

**Comment by elicoidal** [comment]
No company logo or advertisement accepted.
Of course the authoring can/has to mention companies

Eric CAUDAL

Eric Caudal

**Comment by pedrobaeza** [comment]
Hi, one remark without making exhaustive review: first module name must follow conventions about names, as you have made on the seconds, so I propose _account_invoice_supplier_ref_unique_ or _account_invoice_supplier_number_unique_.

Regards.

**Comment by bwrsandman** [comment]
@mcassuto, the sfl logo is a placeholder.
You're supposed to look for some creative commons logos which match the module itself.

**Comment by mcassuto** [comment]
All fixed:
• rename module unique_supplier_invoice_number to account_invoice_unique_supplier_invoice_number
• remove logos
• add .pot file to both modules  account_invoice_unique_supplier_invoice_number and account_invoice_unique_customer_ref
• add fr translation to both modules
• fix typo '# this function only works with one id' in both modules
• fix the missing spaces in the licence text for both modules
• fix the too long lines"

File: account_invoice_unique_supplier_invoice_number/account_invoice.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+        sr_ids = self.search(cr, uid,
+                             [("partner_id", "=", invoice_partner.id)],
+                             context=context)
+        lst = [
+            x.supplier_invoice_number.lower() for x in
+            self.browse(cr, uid, sr_ids, context=context)
+            if x.supplier_invoice_number and x.id not in ids
+        ]
+        if invoice.supplier_invoice_number and invoice.supplier_invoice_number.lower() in lst:
```
Comment (rvalyi) [inline account_invoice_unique_supplier_invoice_number/account_invoice.py]: line is 94 chars, limit will soon be 80 chars in Travis tests
You can beat @bwrsandman if you disagree but this is were OCA stands so far.

File: account_invoice_unique_customer_ref/account_invoice.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
+# -*- encoding: utf-8 -*-
+# #############################################################################
+#
+# OpenERP, Open Source Management Solution
+# This module copyright (C) 2010 - 2014 Savoir-faire Linux
```
Comment (rvalyi) [inline account_invoice_unique_customer_ref/account_invoice.py]: bad indent on the 2 upper lines

**Comment by pedrobaeza** [comment]
Sorry for being too picky, but the name you have chosen can be problematic due to its length: there is a restriction on the size for XML-ID names, and you already catch 47 with this name, what leaves only 17 chars for the rest of the ID. That's why I proposed you the other names. There is also a redundancy in your name, using invoice two times. Following the other module name, why don't you name it _account_invoice_unique_supplier_ref_?

Regards.

**Comment by mcassuto** [comment]
@pedrobaeza : modules renamed
@rvalyi : all files checked and PEP8 complient

**Comment by pedrobaeza** [comment]
One last thing: you have to remove .pyc and .mo files.

Regards.

**Comment by mcassuto** [comment]
@pedrobaeza : done ;-)

**Comment by pedrobaeza** [comment]
Thanks!

:+1:

**Comment by mcassuto** [comment]
@pedrobaeza : just for the understanding, what about having these modules merged ?

**Comment by pedrobaeza** [comment]
We need at least one approval more from another reviewer, so please wait for it. If we see that nobody can review for a while, we can measure its inclussion.

Regards.

**Comment by mcassuto** [comment]
thanks a lot Pedro @pedrobaeza

**Comment by rvalyi** [comment]
@mcassuto 

sorry if that's a dumb question,  but your _check_unique_\* methods work with only the 1st id of the ids collection. What happens if you write several invoices at once, with some API call or via some RPC?

Wouldn't it be better to loop over all ids and enforce the test on all ids? Can you justify why you take only the 1st element?

File: account_invoice_customer_ref_unique/account_invoice.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    def copy(self, cr, uid, ids, default=None, context=None):
+        default = default or {}
+        default.update({
+            'name': '',
+        })
+        return super(AccountInvoice, self).copy(cr, uid, ids, default, context)
+
+    def _check_unique_name_insensitive(self, cr, uid, ids, context=None):
+        # this function only works with one id
+        if ids:
```
Comment (bwrsandman) [inline account_invoice_customer_ref_unique/account_invoice.py]: `if ids and len(ids) == 1:`

File: account_invoice_supplier_ref_unique/account_invoice.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        default = default or {}
+        default.update({
+            'supplier_invoice_number': '',
+        })
+        return super(AccountInvoice, self).copy(cr, uid, ids, default, context)
+
+    def _check_unique_supplier_invoice_number_insensitive(self, cr, uid, ids,
+                                                          context=None):
+        # this function only works with one id
+        if ids:
```
Comment (bwrsandman) [inline account_invoice_supplier_ref_unique/account_invoice.py]: `if ids and len(ids) == 1:`

File: account_invoice_supplier_ref_unique/account_invoice.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            x.supplier_invoice_number.lower() for x in
+            self.browse(cr, uid, sr_ids, context=context)
+            if x.supplier_invoice_number and x.id not in ids
+        ]
+        if (invoice.supplier_invoice_number
+                and invoice.supplier_invoice_number.lower() in lst):
+            return False
+        return True
+
+    msg = _(
```
Comment (bwrsandman) [inline account_invoice_supplier_ref_unique/account_invoice.py]: do here what you did [here](https://github.com/OCA/account-invoicing/pull/2/files#diff-ea02e43001b9562a6ca6660e63620aaeR66)

**Comment by mcassuto** [comment]
@rvalyi : because the id is provided by postgres.... hence it is almost impossible to fall into this case.
We had this discussion with another developer internally and we added this check just to enforce this code robustness.

**Comment by mcassuto** [comment]
@bwrsandman : my bad for the langage translation... I forgot. 
Other comment added

**Comment by rvalyi** [comment]
:+1:  ok thanks for justifying. Accepted.

**Comment by max3903** [comment]
:+1:

File: account_invoice_supplier_ref_unique/account_invoice.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+            self.browse(cr, uid, sr_ids, context=context)
+            if x.supplier_invoice_number and x.id not in ids
+        ]
+        if (invoice.supplier_invoice_number
+                and invoice.supplier_invoice_number.lower() in lst):
+            return False
+        return True
+
+    def _rec_message(self, cr, uid, ids, context=None):
+        return _('TThe supplier invoice number must be unique \
```
Comment (bwrsandman) [inline account_invoice_supplier_ref_unique/account_invoice.py]: > TThe

**Comment by mcassuto** [comment]
@bwrsandman : typo fixed and checked in pot/po files

File: account_invoice_supplier_ref_unique/account_invoice.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+    def _check_unique_supplier_invoice_number_insensitive(self, cr, uid, ids,
+                                                          context=None):
+        # this function only works with one id
+        if ids and len(ids) == 1:
+            i_id = ids[0]
+        else:
+            raise orm.except_orm(
+                _('Error'),
+                'Cannot check unique supplier ref without id.')
```
Comment (bwrsandman) [inline account_invoice_supplier_ref_unique/account_invoice.py:45]: This string isn't translated

File: account_invoice_supplier_ref_unique/account_invoice.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        return True
+
+    def _rec_message(self, cr, uid, ids, context=None):
+        return _('The supplier invoice number must be unique \
+                 for each supplier !')
+
+    _constraints = [
+        (_check_unique_supplier_invoice_number_insensitive,
+         _rec_message,
+         ['name'])
```
Comment (bwrsandman) [inline account_invoice_supplier_ref_unique/account_invoice.py:70]: Unless you add `'supplier_invoice_number'` to the list, the constraint will only be checked when name changes...

File: account_invoice_customer_ref_unique/account_invoice.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+        })
+        return super(AccountInvoice, self).copy(cr, uid, ids, default, context)
+
+    def _check_unique_name_insensitive(self, cr, uid, ids, context=None):
+        # this function only works with one id
+        if ids and len(ids) == 1:
+            i_id = ids[0]
+        else:
+            raise orm.except_orm(_('Error'),
+                                 'Cannot check unique name without id.')
```
Comment (bwrsandman) [inline account_invoice_customer_ref_unique/account_invoice.py:43]: This string isn't translated

**Comment by bwrsandman** [comment]
:-1: 
Branch is in PR against the wrong branch (master)
You will need to close this one and make a new PR with the same branch.
Here's a link to do it quickly: https://github.com/savoirfairelinux/account-invoicing/compare/OCA:7.0...savoirfairelinux:7.0_account_invoice_unique_customer_ref

**Comment by bwrsandman** [comment]
@max3903 @rvalyi @pedrobaeza, please check the target branches before approving.

**Comment by mcassuto** [comment]
Pull request made on the wrong branch

**Comment by rvalyi** [comment]
sorry

**Comment by pedrobaeza** [comment]
Sorry for the mistake, but these modules can also apply to 8.0, doesn't it?

**Pull request closed without merging**